nilearn/surface.py

```diff
@@ -0,0 +1,572 @@
+        space.
+
+    segment_half_width : float, optional (default=3.)
+        size in mm of the neighbourhood around each vertex in which to draw
+        samples
+
+    n_points : int, optional (default=10)
+        number of samples to draw for each vertex.
+
+    Returns
+    -------
+    numpy array, shape (n_vertices, n_points, 3)
+        The locations, in voxel space, from which to draw samples.
+        First dimension iterates over mesh vertices, second dimension iterates
+        over the sample points associated to a vertex, third dimension is x, y,
+        z in voxel space.
+
+    """
+    vertices, faces = mesh
+    normals = _vertex_outer_normals(mesh)
+    offsets = np.linspace(-segment_half_width, segment_half_width, n_points)
+    sample_locations = vertices[
+        np.newaxis, :, :] + normals * offsets[:, np.newaxis, np.newaxis]
+    sample_locations = np.rollaxis(sample_locations, 1)
+    sample_locations_voxel_space = np.asarray(
+        resampling.coord_transform(
+            *np.vstack(sample_locations).T,
+            affine=np.linalg.inv(affine))).T.reshape(sample_locations.shape)
+    return sample_locations_voxel_space
+
+
+def _sample_locations(mesh, affine, radius, kind='line', n_points=None):
+    """Get either ball or line sample locations."""
+    projectors = {
+        'line': _line_sample_locations,
+        'ball': _ball_sample_locations
+    }
+    if kind not in projectors:
+        raise ValueError(
+            '"kind" must be one of {}'.format(tuple(projectors.keys())))
+    projector = projectors[kind]
+    # let the projector choose the default for n_points
+    # (for example a ball probably needs more than a line)
+    loc_kwargs = ({} if n_points is None else {'n_points': n_points})
+    sample_locations = projector(
+        mesh, affine, radius, **loc_kwargs)
+    return sample_locations
+
+
+def _masked_indices(sample_locations, img_shape, mask=None):
+    """Get the indices of sample points which should be ignored.
+
+    Parameters:
+    -----------
+    sample_locations : array, shape(n_sample_locations, 3)
+        The coordinates of candidate interpolation points
+
+    img_shape : tuple
+        The dimensions of the image to be sampled
+
+    mask : array of shape img_shape or None
+        Part of the image to be masked. If None, don't apply any mask.
+
+    Returns
+    -------
+    array of shape (n_sample_locations,)
+        True if this particular location should be ignored (outside of image or
+        masked).
+
+    """
+    masked = (sample_locations < 0).any(axis=1)
+    for dim, size in enumerate(img_shape):
+        masked = np.logical_or(masked, sample_locations[:, dim] >= size)
+    if mask is not None:
+        indices = np.asarray(np.round(sample_locations[~masked]), dtype=int)
+        masked[~masked] = mask[
+            indices[:, 0], indices[:, 1], indices[:, 2]] == 0
+    return masked
+
+
+def projection_matrix(mesh, affine, img_shape,
```

Comment:
Why is it public ? What is the use case of sparse matrix ? Useful only when interpolation='nearest' ?

Reply:
There has been quit a bit of discussion about this between @GaelVaroquaux  and @eickenberg (see above in this thread). Apparently having the projection matrix could be useful for advanced users. I have no problem with making this function private if you think it is better.